### PR TITLE
refactor(cloud-crawler): create website scan result updater

### PR DESCRIPTION
**Pull request by waabid**
#### Description of changes

Creates/refactors the updating of the website scan result for both report and url scenarios.

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Added relevant unit test for your changes. (`yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/web-api-scan-runner/src/runner/url-deduplicator.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { injectable } from 'inversify';
+import { uniq } from 'lodash';
+
+@injectable()
+export class UrlDeduplicator {
```
Comment (RobGallo) [inline packages/web-api-scan-runner/src/runner/url-deduplicator.ts:8]: I don't think this needs to be a class; it maintains no state.

File: packages/web-api-scan-runner/src/runner/url-deduplicator.spec.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    });
+
+    test('duplicates in given urls', () => {
+        const old: string[] = ['abc', 'def'];
+        const newUrls: string[] = ['abc', 'ghi'];
+        const expected: string[] = ['abc', 'def', 'ghi'];
+        expect(testSubject.dedupe(old, newUrls)).toEqual(expected);
+    });
+
+    test('no dupelicates', () => {
```
Comment (RobGallo) [inline packages/web-api-scan-runner/src/runner/url-deduplicator.spec.ts:19]:
```suggestion
    test('no duplicates', () => {
```

File: packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        const knownPages = websiteScanResult.knownPages;
+        const updatedKnownPages = this.urlDeduplicator.dedupe(knownPages, newlyDiscoveredUrls);
+
+        const updatedWebsiteScanResults: Partial<WebsiteScanResult> = {
+            id: websiteScanResult.id,
+            knownPages: updatedKnownPages,
+            _etag: websiteScanResult._etag,
+        };
+
+        return this.updateWebsiteScanResult(updatedWebsiteScanResults);
```
Comment (RobGallo) [inline packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts:119]: It _may_ be better to use `await` here rather then `return`. My understanding is that using await doesn't add much overhead and makes it possible to get a meaningful callstack when debugging.

File: packages/web-api-scan-runner/src/runner/url-deduplicator.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    public dedupe: (discoveredUrls: string[], newUrls: string[]) => string[] = (discoveredUrls, newUrls) => {
```
[the remark this answers is not in the document]
Comment (pownkel) [reply packages/web-api-scan-runner/src/runner/url-deduplicator.ts:9]: Maybe addUniqueUrls is more descriptive?

File: packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                });
+            },
+            this.maxCombinedResultProcessingRetryCount,
+            1000,
+        );
+    }
+
+    private async generateCombinedScanResultsImpl(axeScanResults: AxeScanResults, pageScanResult: OnDemandPageScanResult): Promise<void> {
+        const websiteScanRef = this.getWebsiteScanRef(pageScanResult, 'consolidated-scan-report');
+        if (!websiteScanRef) {
```
[the remark this answers is not in the document]
Comment (pownkel) [reply packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts:63]: I don't think it's important to log this. If there's no websiteScanRef, that just means the scan was submitted without deepScan or consolidated report options.

File: packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+        return this.updateWebsiteScanResult(updatedWebsiteScanResults);
+    }
+
+    private async updateWebsiteScanResult(updatedWebsiteScanResults: Partial<WebsiteScanResult>, description?: string): Promise<void> {
+        try {
+            this.websiteScanResultProvider.mergeOrCreate(updatedWebsiteScanResults);
+            this.logger.logInfo(`Successfully updated website scan results${description}.`);
+        } catch (error) {
+            this.logger.logError(`Failed to update website scan results with combined result metadata${description}.`, {
```
Comment (pownkel) [inline packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts:127]:
```suggestion
            this.logger.logError(`Failed to update website scan results with combined result metadata ${description}.`, {
```

File: packages/web-api-scan-runner/src/runner/url-deduplicator.ts (hunk 1 above)
Comment (waabid) [reply packages/web-api-scan-runner/src/runner/url-deduplicator.ts:8]: I was thinking that as well but I couldn't get it to inject using inversify (not supported I believe). After discussing with Kelsey, the way this is dealt with is by using default params. I think that could potentially lead to other issues in the future but for now, will change it to that.

File: packages/web-api-scan-runner/src/runner/add-unique-urls.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { uniq } from 'lodash';
+
+export const AddUniqueUrls = (discoveredUrls: string[], newUrls: string[]) => {
```
Comment (pownkel) [inline packages/web-api-scan-runner/src/runner/add-unique-urls.ts:6]: nit: addUniqueUrls should be lower camel case

**Review by pownkel — APPROVED** [review]

**Pull request merged**